vms/proposervm/proposer/windower.go

```diff
@@ -148,3 +159,118 @@ func (w *windower) Delay(ctx context.Context, chainHeight, pChainHeight uint64,
 	}
 	return delay, nil
 }
+
+func (w *windower) ExpectedProposer(
+	ctx context.Context,
+	blockHeight,
+	pChainHeight,
+	slot uint64,
+) (ids.NodeID, error) {
+	source := prng.NewMT19937_64()
+	sampler, validators, err := w.makeSampler(ctx, pChainHeight, source)
+	if err != nil {
+		return ids.EmptyNodeID, err
+	}
+
+	return w.expectedProposer(
+		validators,
+		source,
+		sampler,
+		blockHeight,
+		slot,
+	)
+}
+
+func (w *windower) MinDelayForProposer(
+	ctx context.Context,
+	blockHeight,
+	pChainHeight uint64,
+	nodeID ids.NodeID,
+	startSlot uint64,
+) (time.Duration, error) {
+	source := prng.NewMT19937_64()
+	sampler, validators, err := w.makeSampler(ctx, pChainHeight, source)
+	if err != nil {
+		return 0, err
+	}
+
+	maxSlot := startSlot + MaxLookAheadSlots
+	for slot := startSlot; slot < maxSlot; slot++ {
+		expectedNodeID, err := w.expectedProposer(
+			validators,
+			source,
+			sampler,
+			blockHeight,
+			slot,
+		)
+		if err != nil {
+			return 0, err
+		}
+
+		if expectedNodeID == nodeID {
+			return time.Duration(slot) * WindowDuration, nil
+		}
+	}
+
+	// no slots scheduled for the max window we inspect. Return max delay
+	return time.Duration(maxSlot) * WindowDuration, nil
+}
+
+func (w *windower) makeSampler(
+	ctx context.Context,
+	pChainHeight uint64,
+	source sampler.Source,
+) (sampler.WeightedWithoutReplacement, []validatorData, error) {
+	// Get the canconical representation of the validator set at the provided
+	// p-chain height.
+	validatorsMap, err := w.state.GetValidatorSet(ctx, pChainHeight, w.subnetID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	validators := make([]validatorData, 0, len(validatorsMap))
+	for k, v := range validatorsMap {
+		validators = append(validators, validatorData{
+			id:     k,
+			weight: v.Weight,
+		})
+	}
+
+	// Note: validators are sorted by ID. Sorting by weight would not create a
+	// canonically sorted list.
+	utils.Sort(validators)
+
+	weights := make([]uint64, len(validators))
+	for i, validator := range validators {
```

Comment:
If we return `totalWeight` from this function, we can remove some additional logic from `Proposers`

Reply:
We only need to calculate the total weight once in `Proposers` - so I don't really see a reason to do that for every call we make to this function